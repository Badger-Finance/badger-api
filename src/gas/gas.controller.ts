import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { GasService } from './gas.service';
import { GasPrices } from './interfaces/gas-prices.interface';
import { GasPricesModel } from './interfaces/gas-prices-model';

@Controller('/gas')
export class GasController {
  @Inject()
  gasService!: GasService;

  @Get('')
  @ContentType('json')
  @Summary('Get the current gas price')
  @Description('Returns the current gas price on the requested chain')
  @Returns(200, GasPricesModel)
  @Returns(404).Description('Chain gas prices not available')
  async getGasPrices(@QueryParams('chain') chain?: Network): Promise<GasPrices> {
    return this.gasService.getGasPrices(Chain.getChain(chain));
  }
}
