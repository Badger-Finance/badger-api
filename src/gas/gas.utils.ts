import NodeCache from 'node-cache';

import { SUPPORTED_CHAINS } from '../chains/chain';
import { ChainGasPrices } from './interfaces/gas-prices.interface';

const gasCache = new NodeCache({ stdTTL: 15, checkperiod: 17 });

export async function getGasCache(): Promise<ChainGasPrices> {
  const cachedGasPrices = gasCache.get<ChainGasPrices>('gasPrices');
  if (cachedGasPrices) {
    return cachedGasPrices;
  }
  const chainGasPrices = await Promise.all(SUPPORTED_CHAINS.map(async (c) => [c.network, await c.getGasPrices()]));
  const gasPrices = Object.fromEntries(chainGasPrices);
  gasCache.set('gasPrices', gasPrices);
  return gasPrices;
}
