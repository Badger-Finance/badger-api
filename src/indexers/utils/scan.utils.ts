import { Network } from '@badger-dao/sdk';
import { BlocksRangeOptions } from '@badger-dao/sdk/lib/common';

import { getDataMapper } from '../../aws/dynamodb.utils';
import { IndexingMetadata } from '../../aws/models/indexing-metadata.model';
import { MAX_SCAN_RANGE } from '../constants/scan.constants';
import { LastScannedBlockMeta } from '../interfaces/last-scanned-block-meta.interface';

export async function getOrCreateMetadata<T>(task: string): Promise<IndexingMetadata<T>> {
  const mapper = getDataMapper();

  let metadata!: IndexingMetadata<T>;

  try {
    metadata = await mapper.get(Object.assign(new IndexingMetadata<T>(), { task }));
  } catch (e) {
    metadata = Object.assign(new IndexingMetadata<T>(), { task, data: {} });
  }

  return metadata;
}

export function getLastScanedBlock<T extends LastScannedBlockMeta>(
  indexingMetaData: T,
  network: Network,
  address: string,
) {
  return indexingMetaData[network]?.[address]?.lastScannedBlock || 0;
}

export function getScanRangeOpts(
  fromChain: Required<BlocksRangeOptions>,
  lastScannedBlock: number,
): Required<BlocksRangeOptions> {
  const startBlock = lastScannedBlock > 0 ? lastScannedBlock + 1 : fromChain.startBlock;
  let endBlock = fromChain.endBlock;

  if (endBlock - startBlock > MAX_SCAN_RANGE) {
    endBlock = startBlock + MAX_SCAN_RANGE;
  }

  return {
    startBlock,
    endBlock,
  };
}

export async function saveIndexingMetadata<T>(model: IndexingMetadata<T>): Promise<void> {
  const mapper = getDataMapper();

  try {
    await mapper.put(model);
  } catch (e) {
    console.error(`Failed to save indexing metadata for ${model.task}`, e);
  }
}
