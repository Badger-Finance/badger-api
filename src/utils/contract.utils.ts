import { ethers } from 'ethers';
import { RawAbiDefinition } from 'typechain/dist/parser/abiParser';

function getMethodSignature({ inputs, name }: RawAbiDefinition) {
  const params = inputs?.map((input) => input.type).join(',');

  return `${name}(${params})`;
}

function getSignatureHash(signature: string) {
  return ethers.utils.keccak256(ethers.utils.toUtf8Bytes(signature)).toString();
}

function getMethodMeta(method: RawAbiDefinition) {
  const methodSignature = getMethodSignature(method);
  const signatureHash = getSignatureHash(methodSignature);

  return { methodSignature, signatureHash };
}

export function extendAbiMethodsWithMeta(abi: RawAbiDefinition[]) {
  const methods = abi.filter((method) => method.type === 'function');

  return methods.map((method) => ({
    ...getMethodMeta(method),
    ...method,
    decodedParams: [],
  }));
}
