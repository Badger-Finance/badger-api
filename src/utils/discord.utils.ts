import axios from 'axios';
import { DISCORD_WEBHOOK_URL } from '../config/constants';

export function sendErrorToDiscord(e: Error, errorMsg: string, errorType: string) {
  sendMessageToDiscord(
    `**${errorType}**`,
    `:x: ${errorMsg}`,
    [
      {
        name: 'Error Name',
        value: e.name,
        inline: true,
      },
      {
        name: 'Error Description',
        value: e.message,
        inline: true,
      },
      {
        name: 'Error Stack',
        value: e.stack?.toString() ?? '',
        inline: false,
      },
    ],
    'Error Bot',
  );
}

export async function sendMessageToDiscord(
  title: string,
  description: string,
  fields: { name: string; value: string; inline: boolean }[],
  username: string,
  url: string = DISCORD_WEBHOOK_URL,
) {
  try {
    await axios.post(url, {
      embeds: [
        {
          title,
          description,
          fields,
        },
      ],
      username,
    });
  } catch (error) {
    console.log(error);
  }
}

export async function sendPlainTextToDiscord(message: string, username: string, url: string = DISCORD_WEBHOOK_URL) {
  try {
    await axios.post(url, { content: message, username });
  } catch (error) {
    console.log(error);
  }
}

export async function sendCodeBlockToDiscord(message: string, username: string, url: string = DISCORD_WEBHOOK_URL) {
  const msg = '```\n' + message + '\n```';
  try {
    await axios.post(url, { content: msg, username });
  } catch (error) {
    console.log(error);
  }
}
