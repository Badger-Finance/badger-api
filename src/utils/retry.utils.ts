// improve, add list of errors to retry (can be only IO errors, or whatever)
export function retryFuncWrapper<A, R>(func: A, retryTimes: number = 2) {
  if (!func) throw new Error('func is required');
  if (typeof func !== 'function') throw new Error('func must be a function');

  return async function (...args: A extends (...args: infer P) => R ? P : never[]) {
    let retryCount = 0;

    let lastError: Error | unknown = null;

    while (retryCount < retryTimes) {
      try {
        if (func.constructor.name !== 'AsyncFunction') {
          return await func(...args);
        }

        return func(...args);
      } catch (err) {
        retryCount++;

        lastError = err;
      }
    }

    console.error('Retry failed');
    throw lastError;
  };
}

// shortcut
export const rfw = retryFuncWrapper;
