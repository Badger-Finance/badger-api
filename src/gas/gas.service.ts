import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { getGasCache } from './gas.utils';
import { GasPrices } from './interfaces/gas-prices.interface';

@Service()
export class GasService {
  /**
   * Attempt to retrieve the prices for a chain from the cache
   * if no value is in the cache, refresh prices for all chains
   * @param chain Gas price chain
   * @returns object of gas speeds and prices
   */
  async getGasPrices(chain: Chain): Promise<GasPrices> {
    const gasCache = await getGasCache();
    return gasCache[chain.network];
  }
}
