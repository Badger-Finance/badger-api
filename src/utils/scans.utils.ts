import axios from 'axios';
import { RawAbiDefinition } from 'typechain/dist/parser/abiParser';

import { sleep } from './process.utils';

const abiCacheMap: {
  [key: string]: RawAbiDefinition[];
} = {};

export function formScanApiUrl(explorerUrl: string): string {
  const expUrlObj = new URL(explorerUrl);
  return `${expUrlObj.protocol}//api.${expUrlObj.host}/api`;
}

export async function getContractAbi(
  address: string,
  scanApiUrl: string,
  withThreshold = true,
  useCache = true,
): Promise<RawAbiDefinition[] | null> {
  if (useCache && abiCacheMap[address]) {
    return abiCacheMap[address];
  }

  // if we don't use private keys for scans they can block us, so it's better to use sleep
  if (withThreshold) await sleep(500);

  const scanResp = await axios.get<{
    result: string;
  }>(`${scanApiUrl}?module=contract&action=getabi&address=${address}`);

  const scanRespJson = scanResp.data;

  if (scanResp.status !== 200) return null;

  if (!scanRespJson || !scanRespJson?.result) {
    console.warn(`Failed to get abi for ${address}`);
    return null;
  }

  const contractAbi = JSON.parse(scanRespJson.result);

  if (useCache) {
    abiCacheMap[address] = contractAbi;
  }

  return contractAbi;
}
