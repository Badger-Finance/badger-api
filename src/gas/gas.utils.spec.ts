import { Network } from '@badger-dao/sdk';
import * as gasUtils from './gas.utils';

describe('gas.utils', () => {
  beforeEach(() => {
    jest.spyOn(gasUtils, 'getGasCache').mockReturnValue(
      Promise.resolve({
        [Network.Arbitrum]: { rapid: 2, fast: 2, standard: 2, slow: 2 },
        [Network.Ethereum]: {
          rapid: { maxFeePerGas: 223.06, maxPriorityFeePerGas: 3.04 },
          fast: { maxFeePerGas: 221.96, maxPriorityFeePerGas: 1.94 },
          standard: { maxFeePerGas: 221.91, maxPriorityFeePerGas: 1.89 },
          slow: { maxFeePerGas: 221.81, maxPriorityFeePerGas: 1.79 },
        },
        [Network.Polygon]: { rapid: 38, fast: 33, standard: 33, slow: 33 },
        [Network.BinanceSmartChain]: {
          rapid: 5,
          fast: 5,
          standard: 5,
          slow: 5,
        },
      }),
    );
  });

  describe('getGasPrices', () => {
    it('returns gas prices for all networks', async () => {
      const gasPrices = await gasUtils.getGasCache();
      expect(gasPrices).toMatchSnapshot();
    });
  });
});
