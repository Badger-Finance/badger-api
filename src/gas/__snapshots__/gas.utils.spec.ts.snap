// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gas.utils getGasPrices returns gas prices for all networks 1`] = `
Object {
  "arbitrum": Object {
    "fast": 33,
    "rapid": 38,
    "slow": 33,
    "standard": 33,
  },
  "avalanche": Object {
    "fast": 33,
    "rapid": 38,
    "slow": 33,
    "standard": 33,
  },
  "binancesmartchain": Object {
    "fast": 33,
    "rapid": 38,
    "slow": 33,
    "standard": 33,
  },
  "ethereum": Object {
    "fast": Object {
      "maxFeePerGas": 221.96,
      "maxPriorityFeePerGas": 1.94,
    },
    "rapid": Object {
      "maxFeePerGas": 223.06,
      "maxPriorityFeePerGas": 3.04,
    },
    "slow": Object {
      "maxFeePerGas": 221.81,
      "maxPriorityFeePerGas": 1.79,
    },
    "standard": Object {
      "maxFeePerGas": 221.91,
      "maxPriorityFeePerGas": 1.89,
    },
  },
  "fantom": Object {
    "fast": 536.4688,
    "rapid": 536.4688,
    "slow": 536.4688,
    "standard": 536.4688,
  },
  "polygon": Object {
    "fast": 33,
    "rapid": 38,
    "slow": 33,
    "standard": 33,
  },
}
`;
